### ISSUE TITLE
FFI: Add vector/SIMD operations

### ISSUE THREAD

**Issue by MikePall** [LuaJIT/LuaJIT#40]
Currently, vector data types may be defined with the FFI, but you really can't do much with them. The goal of this project is to add full support for vector data types to the JIT compiler and the CPU-specific backends (if the target CPU has a vector extension).

A new `ffi.vec` module should declare standard vector types and attaches the machine-specific SIMD intrinsics as (meta)methods.

Prerequisites for this project are user-definable intrinsics #39 and the new garbage collector #38 .

**Comment by corsix** [comment]
> and the new garbage collector #38

Am I correct in assuming that this prerequisite is because vector data types typically require alignment greater than 8 bytes, and that the new garbage collector (or, more precisely, the new allocator) would make such aligned allocations relatively cheap (and even free for 16 byte alignment)?

**Comment by MikePall** [comment]
Yes. The FFI cdata alignment handling for the current GC is an ugly hack. Basically, treating it as a variable-sized allocation and prepending enough of a header to get the desired alignment. That's an enormous waste for a (single) vector cdata object.
